lib/srv/db/common/auth.go

```diff
@@ -293,6 +308,45 @@ func (a *dbAuth) GetAzureAccessToken(ctx context.Context, sessionCtx *Session) (
 	return token.Token, nil
 }
 
+// GetAzureCacheForRedisToken retrieves auth token for Azure Cache for Redis.
+func (a *dbAuth) GetAzureCacheForRedisToken(ctx context.Context, sessionCtx *Session) (string, error) {
+	resourceID, err := arm.ParseResourceID(sessionCtx.Database.GetAzure().ResourceID)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	var client libazure.CacheForRedisClient
+	switch resourceID.ResourceType.String() {
+	case "Microsoft.Cache/Redis":
+		client, err = a.cfg.Clients.GetAzureRedisClient(resourceID.SubscriptionID)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+	case "Microsoft.Cache/redisEnterprise":
+		client, err = a.cfg.Clients.GetAzureRedisEnterpriseClient(resourceID.SubscriptionID)
+		if err != nil {
+			return "", trace.Wrap(err)
+		}
+	default:
+		return "", trace.BadParameter("unknown Azure Cache for Redis resource type: %v", resourceID.ResourceType)
+	}
+
+	token, err := client.GetToken(ctx, resourceID.ResourceGroupName, resourceID.Name)
+	if err != nil {
+		if trace.IsAccessDenied(err) {
+			// Some Azure error messages are long, multi-lined, and may even
+			// contain divider lines like "------". It's unreadable in
+			// redis-cli as the message has to be merged to a single line
+			// string. Thus logging the original error as debug and returning a
+			// more user friendly message.
+			a.cfg.Log.WithError(err).Debugf("Failed to get token for Azure Redis %q.", sessionCtx.Database.GetName())
+			return "", trace.AccessDenied("Failed to get token for Azure Redis %q. Please make sure the database agent has the \"listKeys\" permission to the database.", sessionCtx.Database.GetName())
+		}
+		return "", trace.Wrap(err)
+	}
```

Comment:
Could we add a check for trace.NotFound here, since that could be returned by redis api `ListKeys`? It could return that in the converted azcore response error if the resource group or name are not found.

And maybe it would be best to always log the original error to debug, but
return either `trace.AccessDenied(...)`, `trace.NotFound(...)` or `trace.Errorf` with a simpler message:

```suggestion
	if err != nil {
		// Some Azure error messages are long, multi-lined, and may even
		// contain divider lines like "------". It's unreadable in
		// redis-cli as the message has to be merged to a single line
		// string. Thus logging the original error as debug and returning a
		// more user friendly message.
		a.cfg.Log.WithError(err).Debugf("Failed to get token for Azure Redis %q.", sessionCtx.Database.GetName())
		if trace.IsAccessDenied(err) {
			return "", trace.AccessDenied("Failed to get token for Azure Redis %q. Please make sure the database agent has the \"listKeys\" permission to the database.", sessionCtx.Database.GetName())
		}
		if trace.IsNotFound(err) {
			return "", trace.NotFound("Failed to get token for Azure Redis %q. Please make sure the database resource ID is correct.", sessionCtx.Database.GetName())
		}
		return "", trace.Errorf("Failed to get token for Azure Redis %q.", sessionCtx.Database.GetName())
	}

```

Reply:
Good catch. Will fix.